Expanding tests and docs relating to deleting profiles (#8351)
NixOS/nix: doc/manual/src/command-ref/nix-collect-garbage.md

# Name

`nix-collect-garbage` - delete unreachable [store objects]

# Synopsis

`nix-collect-garbage` [`--delete-old`] [`-d`] [`--delete-older-than` *period*] [`--max-freed` *bytes*] [`--dry-run`]

# Description

The command `nix-collect-garbage` is mostly an alias of [`nix-store --gc`](@docroot@/command-ref/nix-store/gc.md),
that is, `nix-collect-garbage` deletes all unreachable [store objects] in the Nix store to clean up your system.

However, it provides two additional options which also delete old [profiles], allowing potentially more [store objects] to be deleted because profiles are also garbage collection roots.
These options are the equivalent of running variations of [`nix-env --delete-generations`](@docroot@/command-ref/nix-env/delete-generations.md)

> on various profiles, prior to running `nix-collect-garbage` (or just `nix-store --gc`) without any flags.

Author: I mentioned "variations" because the argument to that one is different depending on the flag/argument to this command.

Perhaps there is a better way to say this, like "running <cmd> with various arguments"?

> These options are the equivalent of running variations of [`nix-env --delete-generations`](@docroot@/command-ref/nix-env/delete-generations.md)
> on various profiles, prior to running `nix-collect-garbage` (or just `nix-store --gc`) without any flags.

Author: ```suggestion
These options are the equivalent of running
[`nix-env --delete-generations`](@docroot@/command-ref/nix-env/delete-generations.md)
with various augments on multiple profiles,
prior to running `nix-collect-garbage` (or just `nix-store --gc`) without any flags.
```

Do note that deleting previous configurations makes rollbacks to them impossible.

> Because these flags potentially delete older configurations used by other users on the system, these flags should be used with care.

Reviewer: I see. This is a bit subtle and seems hard to express concisely. I'm not even sure if we should expand on "how it was done up til recently" in too much detail. That won't age well and just produce maintenance later on.

In this case, maybe something like "...in legacy locations".

> Because these flags potentially delete older configurations used by other users on the system, these flags should be used with care.

Author: The global profiles (something I try to enshrine better on the next PR) will continue to be impacted, and they are intentionally shared between users, so I think the vague thing I originally wrote is still true even if we phase out the extra compat stuff?

> Because these flags potentially delete older configurations used by other users on the system, these flags should be used with care.

Author: ```suggestion
These flags should be used with care, because they potentially delete generations of profiles used by other users on the system.
```

## Locations searched for profiles

`nix-collect-garbage` cannot know about all profiles; that information doesn't exist.

> Instead, it looks looks in a few locations for profiles, and acts on all profiles it finds there.

Reviewer: ```suggestion
Instead, it looks looks in a few locations, and acts on all profiles it finds there:
```

Firstly, `nix-collect-garbage` looks in the default profile locations as specified in the [profiles] section of the manual.

Secondly, `nix-collect-garbage` looks in `$NIX_STATE_DIR/profiles` and `$NIX_STATE_DIR/profiles/per-user`.
With the exception of `$NIX_STATE_DIR/profiles/per-user/root` and `$NIX_STATE_DIR/profiles/default`, members of these directories are no longer used by default.

> `nix-collect-garbage` look there anyways in order to clean up profiles from older versions of Nix.

Reviewer 2: I'm very much not keen on documenting that. These locations aren't something that people should care about and exposing them just increases our API surface for no good reason. People relying on `/nix/var/nix/profiles` is already what caused most of the troubles with https://github.com/NixOS/nix/pull/5226 (and similarly, https://github.com/NixOS/nix/pull/5588 is blocked by the fact that we expose more files than we should)

Reviewer: We still have to answer the general question how to deal with such issues: There are some things that are accessible, i.e. are already part of the API, but should not be relied on.

Author: I added a note, but not marking this resolve yet until you both approve.

Reviewer 2: Looks good, thanks :+1:

> Firstly, `nix-collect-garbage` looks in the default profile locations as specified in the [profiles] section of the manual.
> 
> Secondly, `nix-collect-garbage` looks in `$NIX_STATE_DIR/profiles` and `$NIX_STATE_DIR/profiles/per-user`.
> With the exception of `$NIX_STATE_DIR/profiles/per-user/root` and `$NIX_STATE_DIR/profiles/default`, members of these directories are no longer used by default.
> `nix-collect-garbage` look there anyways in order to clean up profiles from older versions of Nix.
> This preserves this command historical behavior, when those locations were used by default.

Reviewer: ```suggestion
1. The default profile locations as specified in the [profiles] section of the manual.

2. `$NIX_STATE_DIR/profiles` and `$NIX_STATE_DIR/profiles/per-user`.

   With the exception of `$NIX_STATE_DIR/profiles/per-user/root` and `$NIX_STATE_DIR/profiles/default`, these directories are no longer used by other commands.
   `nix-collect-garbage` looks there anyways in order to clean up profiles from older versions of Nix.
```

The profiles page is not accessible on the manpages, but we could instead include it here and link and refer to that section.

# Options

These options are for deleting old [profiles] prior to deleting unreachable [store objects].

- <span id="opt-delete-old">[`--delete-old`](#opt-delete-old)</span> / `-d`\
  This deletes all old generations of profiles.

  This is the equivalent of invoking `nix-env --delete-generations old` on each found profile.

- <span id="opt-delete-older-than">[`--delete-older-than`](#opt-delete-older-than)</span> *period*\
  This deletes all generations of profiles older than the specified amount (except for the generations that were active at that point in time).

>   *period* is a value such as `30d`, which would mean 30 days.

Author: I did it below when I mentioned the command.

# Name

`nix-collect-garbage` - delete unreachable [store objects]

# Synopsis

`nix-collect-garbage` [`--delete-old`] [`-d`] [`--delete-older-than` *period*] [`--max-freed` *bytes*] [`--dry-run`]

# Description

> The command `nix-collect-garbage` is mostly an alias of [`nix-store --gc`](@docroot@/command-ref/nix-store/gc.md),
> that is, `nix-collect-garbage` deletes all unreachable [store objects] in the Nix store to clean up your system.

Reviewer: ```suggestion
The command `nix-collect-garbage` is mostly an alias of [`nix-store --gc`](@docroot@/command-ref/nix-store/gc.md).
That is, it deletes all unreachable [store objects] in the Nix store to clean up your system.
```

However, it provides two additional options,
[`--delete-old`](#opt-delete-old) and [`--delete-older-than`](#opt-delete-older-than),
which also delete old [profiles], allowing potentially more [store objects] to be deleted because profiles are also garbage collection roots.
These options are the equivalent of running
[`nix-env --delete-generations`](@docroot@/command-ref/nix-env/delete-generations.md)
with various augments on multiple profiles,
prior to running `nix-collect-garbage` (or just `nix-store --gc`) without any flags.

> **Note**
>
> Deleting previous configurations makes rollbacks to them impossible.

Because these flags potentially delete older configurations used by other users on the system, these flags should be used with care.

## Locations searched for profiles

`nix-collect-garbage` cannot know about all profiles; that information doesn't exist.

> Instead, it looks looks in a few locations, and acts on all profiles it finds there:

Reviewer: ```suggestion
Instead, it looks in a few locations, and acts on all profiles it finds there:
```

1. The default profile locations as specified in the [profiles] section of the manual.

2. > **NOTE**
   >
   > Not stable; subject to change
   >
   > Do not rely on this functionality; it just exists for migration purposes and is may change in the future.
   > These deprecated paths remain a private implementation detail of Nix.

   `$NIX_STATE_DIR/profiles` and `$NIX_STATE_DIR/profiles/per-user`.

   With the exception of `$NIX_STATE_DIR/profiles/per-user/root` and `$NIX_STATE_DIR/profiles/default`, these directories are no longer used by other commands.
   `nix-collect-garbage` looks there anyways in order to clean up profiles from older versions of Nix.

# Options

These options are for deleting old [profiles] prior to deleting unreachable [store objects].

- <span id="opt-delete-old">[`--delete-old`](#opt-delete-old)</span> / `-d`\

>   This deletes all old generations of profiles.

Reviewer: ```suggestion
  Delete all old generations of profiles.
```

  This is the equivalent of invoking `nix-env --delete-generations old` on each found profile.

- <span id="opt-delete-older-than">[`--delete-older-than`](#opt-delete-older-than)</span> *period*\

>   This deletes all generations of profiles older than the specified amount (except for the generations that were active at that point in time).

Reviewer: ```suggestion
  Delete all generations of profiles older than the specified amount (except for the generations that were active at that point in time).
```

What does it mean to be active though? Which point in time exactly? Time of running the command, or falling into that interval between then and now?
